docs/MATCHMAKING_PROPOSAL.md

```diff
@@ -0,0 +1,171 @@
+
+The game document `/game/{gameId}` looks like this:
+```jsonc
+// game.json
+
+{
+  "mode": "5P,PC,3R,0T",
+  "timestamp": "2020-01-01T00:00:00.000Z",
+  "rounds": [],
+  "players": [/*players*/],
+  "done": false,
+  "users_state": {
+    "user1": "ABSENT",
+    "user2": "READY",
+    "user3": "JOINED,
+  }
+}
+```
+
+In the game document, the users can have 3 different states:
+* `ABSENT`: the user is not in the game
+* `JOINED`: the user has joined the game, but is not ready yet
+* `READY`: the user has joined the game and is ready to start the game
+
+### 3. Game start - client-side
+
+Now that we have a game id, we can retrieve the game document:
+```
+/games/{gameId}
+```
+and listen to the game document for changes.
+### 4. Ready check (Optional)
+To make sure players are ready, each player needs to press the READY button.
+This will update the `users_state` map of the game document to `READY` for the player. (The player's uid is the key.)
+
+If after a certain amount of time, the game is not ready, the game is cancelled:
+- The document in the queue is marked as done by setting the `done` field to `true`.
+- The game document is marked as done by setting the `done` field to `true`.
+
+### 5. Round
+Once every player is ready, server adds a round document in the `rounds/` collection of the game document `/games/{gameId}/rounds/{roundId}`:
+```jsonc
+// round.json
+
+{
+  "uid": "round_uid",
+  "timestamp": "2020-01-01T00:00:00.000Z",
+  "players": [/*players*/],
+  "game": "game_uid",
+  "hands": {}
+}
+```
+It also adds the round's `id` to the `rounds` field of the game document.
+
+Each player pick their hand (`ROCK`, `PAPER` or `SCISSORS`). The client update the `hands` map of the round document to reflect the player's hand. Say `user1` picked `ROCK`, `user2` picked `PAPER` and `user3` picked `SCISSORS`:
+
+```jsonc
+// round.json
+
+{
+  "uid": "round_uid",
+  "timestamp": "2020-01-01T00:00:00.000Z",
+  "players": [/*players*/],
+  "game": "game_uid",
+  "hands": {
+    "user1": "ROCK",
+    "user2": "PAPER",
+    "user3": "SCISSORS"
+  }
+}
+```
+
+A round is over when each player in `players` has a value in the `hands` map.
+
+The winner of a round is determined by comparing the hands of the players. The winner is the player with the highest score. It is computed like so:
+
+1. Every player gets a score of 0 at the beginning of the round.
+2. Each player is confronted to every other player.
+    - If the player's hand **wins** against the other player's hand, the player **gets a point**.
+    - If the player's hand **loses** against the other player's hand, the player **loses a point**.
+    - If the player's hand **ties** with the other player's hand, **no points are awarded**.
+
+
+### 6. Next Round
+Once a round is over, the server adds a new round document in the `rounds/` collection of the game document `/games/{gameId}/rounds/{roundId}`:
+
+### 7. Game Over
+Once all rounds are over, the game is over.
```

Comment:
Actually, I would say that the user decides on the score of the best user that ends a game. So we stop when we reach the score not the number of rounds

Reply:
How is it better ?

I don't think that's a good idea. Number of rounds are a good idea of how long the game will last, the score to reach isn't. Theoretically speaking, if you are 3 in the lobby and everyone wins 1 every 3 turns, it could take `3*n-2` rounds to reach the top score where `n` the maximum score. So a factor of 3 on the estimated time of a game.

The number of rounds as the constraints is better, as it's a direct and 1 to 1 estimation to the length of the game.

If I only have 5 minutes to play a game, I don't want to have to think whether I can take the risk for my game to reach 15 miutes, or take a shorter game etc.

Reply:
That doesn't reflect a real-life situation (at least the rules I played by). People wouldn't want new rules.

Reply:
This isn't new rules, it's a game format. Sure, a target score is usually chosen when people play in person together, but it's almost never the case in online games AFAIK.

Mainly for the reasons evoked in my previous message.

Also note, that rules you play by don't always make sense, are balanced, or are scalable to online games ;)

Reply:
I think we can give two possibilities and start with the support for the one you're proposing (I'd be really happy to work on the other version later on as a part of my firebase learning). Though, I think that getting rid of the default format/rules/variant is suboptimal.